core/modules/file/file.module

```diff
@@ -2903,19 +2903,28 @@ function file_file_access($op, $file, $account) {
     }
 
     if ($op == 'download') {
-      if (user_access('download any ' . $type . ' files', $account) || is_object($file) && user_access('download own ' . $type . ' files', $account) && ($account->uid == $file->uid)) {
+      if (user_access("download any $type files", $account)) {
+        return FILE_ACCESS_ALLOW;
+      }
+      if ($account->uid && $account->uid == $file->uid && user_access("download own $type files", $account)) {
```

Comment:
The check  of `$account->uid` will fail when dealing with an anonymous user (uid == 0). 

Notice that file permissions DO allow anonymous users to edit, delete and download their "own" files. I suggest you use something like `isset($account->uid)` of perhaps `($account->uid || user_is_anonymous())`

Reply:
BTW, this also applies for edit and delete below.

Reply:
> The check of $account->uid will fail when dealing with an anonymous user (uid == 0).

That's exactly the purpose of this check! Not that it's only done with checks on "own stuff".

Anonymous can not "own" anything. Think of the following: an anonymous user tries to access a file. That file might also have uid = 0 (user isn't in B anymore), then *anyone* (any visitor) would be granted access, which is definitely not what we want.

Reply:
@argiepiano I'm taking the liberty to mark this as resolved - I think my answer explained why this is needed.